### ISSUE TITLE
`Self` is not working correctly with a custom descriptor which captures self type

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#6428]
**Describe the bug**

See sample code below. It used to work with Pyright and is working with Mypy.

**Code or Screenshots**

```python
from __future__ import annotations

from typing import Callable, Generic, Self, TypeVar, overload

T = TypeVar("T", contravariant=True)
R = TypeVar("R", covariant=True)


class MyProperty(Generic[T, R]):
    def __init__(self, func: Callable[[T], R], /) -> None:
        self.__func = func

    @overload
    def __get__(self, instance: None, type_: type[T], /) -> Self:
        ...

    @overload
    def __get__(self, instance: T, type_: type[T] | None = ..., /) -> R:
        ...

    def __get__(self, instance: T | None, type_: type[T] | None = None, /) -> Self | R:
        if instance is None:
            return self
        return self.__func(instance)


class C:
    @MyProperty
    def v(self, /) -> int:
        return 0

    def f(self, /) -> None:
        _ = self.v  # Failing.


_ = C().v  # OK.
```

![image](https://github.com/microsoft/pyright/assets/15389744/54450066-8706-418d-bdeb-c1b494b80576)

**VS Code extension or command-line**
Pylance v2023.11.12 Pre-Release
Pyright 1.1.335

**Comment by erictraut** [comment]
Thanks for the bug report. This was due to the same bug reported in #6452. It will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
